src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -269,6 +269,21 @@ public async ValueTask<IEnumerable<SourceGeneratedDocument>> GetSourceGeneratedD
             return builder.ToImmutable();
         }
 
+        public async ValueTask<SourceGeneratedDocument?> GetSourceGeneratedDocumentAsync(DocumentId documentId, CancellationToken cancellationToken = default)
+        {
+            var generatedDocumentStates = await _solution.State.GetSourceGeneratedDocumentStatesAsync(this.State, cancellationToken).ConfigureAwait(false);
+
+            foreach (var generatedDocumentState in generatedDocumentStates)
+            {
+                if (generatedDocumentState.Id == documentId)
+                {
+                    return ImmutableHashMapExtensions.GetOrAdd(ref _idToSourceGeneratedDocumentMap, generatedDocumentState.Id, s_createSourceGeneratedDocumentFunction, (generatedDocumentState, this));
+                }
+            }
```

Comment:
can we first lookup in _idToSourceGeneratedDocumentMap and avoid the linear walk (and call to get teh states) if the doc is already in the map.

Reply:
Ah yes, of course.